# Part 1

## Pandas - Working with Data

In the section, we discussed using external libraries. This section covers one of the most extensively used libraries - Pandas. Most of the data manipulation you've done so far for visualizations required importing data in Excel. Pandas library makes working with data a breeze. You will find some exceptionally useful functions that make data  - cleaning, exploring, manipulation and generating visualization-ready datasets - easy.

In [2]:
# let's get started.
# note how we use as pd while importing. This is used for assigning a nickname since we will be using library name a lot.
# You can nickname it anything but since its a convention, we will go with pd
import pandas as pd
import numpy as np

In [3]:
df= pd.DataFrame({'A':[np.nan,2,3,np.nan], 'B':['Faizan', 'Hakim', 'Karam', 'Ali']})
df

A       B
0  NaN  Faizan
1  2.0   Hakim
2  3.0   Karam
3  NaN     Ali

In [6]:
df = pd.DataFrame({'secrion A':[ "ayesha", "ayan malik", "talha Amjad", "sabah ahmed"], 'marks':[525, 505, 450, 446]})
df

secrion A  marks
0       ayesha    525
1   ayan malik    505
2  talha Amjad    450
3  sabah ahmed    446

In [7]:
df1 = pd.DataFrame({'C':[1,2,"Jam"]})
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   C       3 non-null      object
dtypes: object(1)
memory usage: 156.0+ bytes


In [79]:
df1 = df1.drop(index=[2], axis=0)

In [80]:
df1

C
0  1
1  2

In [72]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   A       2 non-null      float64
 1   B       4 non-null      object 
dtypes: float64(1), object(1)
memory usage: 192.0+ bytes


In [73]:
df['A'].fillna(1.0, inplace=True)

In [74]:
df

A       B
0  1.0  Faizan
1  2.0   Hakim
2  3.0   Karam
3  1.0     Ali

In [25]:
# First, we explore a new data type: DataFrame!
# data frames take data in any iterable. Let's start with a simple list
mydata = [1,2,3,4,5,6,7,8,9,10]
my_data2 = [11,12,13,14,15,16,17,18,19,20]
df = pd.DataFrame(list(zip(mydata, my_data2)), columns=['Numbers', 'Number2'])
df

Numbers  Number2
0        1       11
1        2       12
2        3       13
3        4       14
4        5       15
5        6       16
6        7       17
7        8       18
8        9       19
9       10       20

In [ ]:
#Data Frame with Dictionary
number_dictionary = {'Numbers':mydata, 'Number2': mydata}
data2 = pd.DataFrame(number_dictionary)
data2

Numbers  Number2
0        1        1
1        2        2
2        3        3
3        4        4
4        5        5
5        6        6
6        7        7
7        8        8
8        9        9
9       10       10

In [2]:
import Data frame as df

SyntaxError: invalid syntax (1608500970.py, line 1)

In [1]:
# let's have a look at our df
print(df)
# note the index on the left: it specifies at what index a value exists

NameError: name 'df' is not defined

In [ ]:
# if we want to view just the top few values of our df, we can use .head() function
df.head(15)

Numbers  Number2
0        1        1
1        2        2
2        3        3
3        4        4
4        5        5
5        6        6
6        7        7
7        8        8
8        9        9
9       10       10

In [ ]:
# similarly, to view the last values, we can use .tail(). By default, both head and tail display 5 values.
# You can pass an argument to these functions to display as many values as you would like. Let's view the last 2

df.tail(6)

Numbers  Number2
4        5        5
5        6        6
6        7        7
7        8        8
8        9        9
9       10       10

In [ ]:
# you can also access specific columns using its name. Here we defined a column named numbers. Let's look at that
df['Number2']

0    11
1    12
2    13
3    14
4    15
5    16
6    17
7    18
8    19
9    20
Name: Number2, dtype: int64

You can also use numeric index values or index numbers to access specific rows. For that we use .loc[] or .iloc[]

For loc, you provide values;

`df.loc[starting index value: ending index value, starting column value:ending column value]`

For example, to access Numbers column's first 2 values, we write:

`df.loc[0:2, 'Numbers']`

Whereas, for iloc, you provide numeric values;
`df.loc[starting index: ending index, starting column index:ending column index]`

To replicate the above example with iloc, we write:

`df.iloc[0:2, 0]`

Note that since index values are the same (default numbering), iloc takes the same row values. However, that won't always be the case. Sometimes there will be dates or other non-numeric labels. It will then make sense why we have two different ways of accessing data i.e. by actual value or by index.

In [27]:
# give the above code a try here. Access the middle 2 values of the Numbers column using iloc and loc.

### start code here

print(df.loc[df.Numbers == 7])
print(df.iloc[df.Nmubers == 8])


### end code here

   Numbers  Number2
6        7       17


AttributeError: 'DataFrame' object has no attribute 'Nmubers'

In [3]:
name = ["Luke", "Anakin", "Ahsoka", "Liea", "Han"]
overall_perct = [55, 65, 75, 80, 95]
grade = ['D', 'C', 'B', 'A', 'A+']

In [6]:
df = pd.DataFrame(list(zip(name, overall_perct, grade)), columns = ["Name", "Overall Percentage", "Grade"])
df

Name  Overall Percentage Grade
0    Luke                  55     D
1  Anakin                  65     C
2  Ahsoka                  75     B
3    Liea                  80     A
4     Han                  95    A+

In [ ]:
df3 = pd.DataFrame()
col_name = ["Name", "Overall Percentage", "Grade"]
values = [["Luke", "Anakin", "Ahsoka", "Liea", "Han"], [55, 65, 75, 80, 95], ['D', 'C', 'B', 'A', 'A+']]
for i in range(len(values)):
  df3[col_name[i]] = values[i]

df3

Name  Overall Percentage Grade
0    Luke                  55     D
1  Anakin                  65     C
2  Ahsoka                  75     B
3    Liea                  80     A
4     Han                  95    A+

In [22]:
b = df['Overall Percentage']
print(b)
for i in b:
 s


0    55
1    65
2    75
3    80
4    95
Name: Overall Percentage, dtype: int64
1
1
1
0
1


In [ ]:
df2 = pd.DataFrame({"Name":name, "Overall Percentage":overall_perct, "Grade":grade})
df2

Name  Overall Percentage Grade
0    Luke                  55     D
1  Anakin                  65     C
2  Ahsoka                  75     B
3    Liea                  80     A
4     Han                  95    A+

In [23]:
#df.iloc[2:4]
df.iloc[2:4, 0:3]
df.iloc[2:4, [0,2]]

Name Grade
2  Ahsoka     B
3    Liea     A

This all looks good and simple so far. How about we take it up a notch and do some calculations on our own dataset?

### Part 1.1 Reading Data into Python Using Pandas

Previously, we used the open() function to read and write from Python. However, pandas let's us read and write files as well however in a much more convienient way. The function read_csv() let's you read any csv file into Python.

We will be using the Forbes 2021 Billionaire Dataset for this section to learn more about pandas.

In [81]:
# let's first define our path
# path = ''
#path = 'C:/Users/LAPTOP PLANET/Documents/Kamran KArachi AI Wor/'
filename = 'Billionaire.csv'
complete_filepath =filename
data = pd.read_csv(filename)
type(data)
#data.head()

pandas.core.frame.DataFrame

In [82]:
pd.set_option('display.max_rows', len(data))

Tip: If you need to know what arguments a function takes, just write it inside help(func_name) without the () of the function call.
For example, if you want to know what arguments you can use with read_csv(), run `help(pd.read_csv)`

In [ ]:
# get help on a function
help(pd.read_csv)

### Part 1.2 Data Exploration

In [66]:
# the data variable is of type dataframe. This means we can use the .head() function to view top rows of our data.
data.head()

Name NetWorth        Country         Source  Rank
0                Jeff Bezos   $177 B  United States         Amazon     1
1                 Elon Musk   $151 B  United States  Tesla, SpaceX     2
2  Bernard Arnault & family   $150 B         France           LVMH     3
3                Bill Gates   $124 B  United States      Microsoft     4
4           Mark Zuckerberg    $97 B  United States       Facebook     5

In [83]:
# you can also look at the number of rows and columns using the shape function.
# It returns a 2x1 array (numberofrows, numberofcolumns)
data.shape

(2755, 5)

In [ ]:
# to view available columns in our dataset, we use the .columns attribute.
# Attribute are properties and do not require () at the end, like functions do.

data.columns

Index(['Name', 'NetWorth', 'Country', 'Source', 'Rank'], dtype='object')

#### Questions to Ask About Data

So we have names, the networth, the country of origin, source of income and ranking for all these billionaires. When you have any datasets, there are some fundamental questions you would want to ask yourself and then find those answers within the data. For instance, for our dataset we can ask ourselves:
1. How many billionaires are there in the list?
2. Who are the top 10 billionaire?
3. Who are the top billionaire by country of origin?
4. What country has the highest number of billionaires?
5. What country has the least number of billionaires?
6. What is the average and median networth of each billionaire?
7. Are there any outliers i.e. billionaires that have exceptionally high networth compared to others?
8. For each country, what is the cumulative networth of its billionaires?
9. What is the maximum and minimum networth of billionaires overall and by country?
10. Are there are any billionaires with the same source of income?

#### But First, Data Cleaning

Before we dive into answering these questions, we need to make sure of the following (and this applies to all datasets that you will encounter in future so keep this checklist in mind):
1. are there any null or missing values in the data?
2. are the data types of each column what you are expecting it to be?

Let's tackle these questions one by one.

A1. Missing values in Python or most datasets in Python are represented by nan or null. Pandas comes with various options to fill missing values, drop them or simply replace them.

*Note: Sometimes, curated datasets may use a special indicator for missing values such as -99999 or some large value like that. Similar to the replace() function we can use .replace() on a pandas dataframe and replace such missing values placeholders with nan and then use the missing value handling methods on the dataset.*

#### Handling Missing Values

In [ ]:
# let's have a look whether there are any missing values in here
# this function checks value by value if it is nan or not. If it is, it returns a True, otherwise a False.
# the function itself returns a new dataframe (the actual data is safe!) with Falses and Trues accordingly
data.isna()

Each False here represents that the value is indeed not a nan or missing value. However, it is barely possible to go through all 2755x5 values. So we can use a summarize function on top of this called `.any()`. Any returns False if **any** of the value here was a True and returns a `True` otherwise `False` (similar to how an OR() function works in Excel) whereas, a function called `.all()` checks if all values are True and hence returns a `True`, otherwise `False` (similar to AND function in Excel)

In [ ]:
# Since the returned data type is also a dataframe (you can check using the type() method), you can use .any() on it again.
data.isna().any()

Name        False
NetWorth     True
Country      True
Source       True
Rank        False
dtype: bool

Great. So there are no missing values in our dataset. What if there had been how would we handle that?
This requires a bit of thought process on your part. since we have individual billionaires, and if there were missing values say in country of origin, you would have to manually search and fill those spaces since no mathematical option (such as using averages) would make sense. Or you could ideally just drop those values specially if their ranks are in the lower half of the dataset for instance. To do this, you can use .dropna(). By default, it drops rows with missing values. However, if you want to drop columns with missing values use .dropna(axis=1).

In [ ]:
# since our data does not have any nan values, using dropna() makes no difference.
# we see no difference is made since the data is still the same shape
dropped_data = data.dropna(inplace=False)
print(dropped_data.shape)

(2755, 5)


In [ ]:
dropped_data.isna().any()

Name        False
NetWorth    False
Country     False
Source      False
Rank        False
dtype: bool

In [ ]:
# to check if both the dropped_data and data dataframes are the same, we use another method called .equals()
# this function takes another dataframe as an argument and returns true if both the dataframes are equal
data.equals(dropped_data)

False

#### Handling Data Types

We need to ensure that data has the right type to perform calculations on. There are two columns in our dataset that are numeric i.e. `Rank` and `Networth`. Dataframes have a cool function called `.info()` that gives some preliminary info on data types and nulls within a dataset on column-wise basis. Let's have a look at the info of our data.

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2755 entries, 0 to 2754
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Name      2755 non-null   object
 1   NetWorth  2753 non-null   object
 2   Country   2751 non-null   object
 3   Source    2753 non-null   object
 4   Rank      2755 non-null   object
dtypes: object(5)
memory usage: 107.7+ KB


In [ ]:
data

Name NetWorth        Country         Source  Rank
0                   Jeff Bezos   $177 B  United States         Amazon     1
1                    Elon Musk   $151 B  United States  Tesla, SpaceX     2
2     Bernard Arnault & family   $150 B         France           LVMH     3
3                   Bill Gates   $124 B  United States      Microsoft     4
4              Mark Zuckerberg    $97 B  United States       Facebook     5
...                        ...      ...            ...            ...   ...
2750         Daniel Yong Zhang     $1 B          China     e-commerce  2674
2751             Zhang Yuqiang     $1 B          China     Fiberglass  2674
2752             Zhao Meiguang     $1 B          China    gold mining  2674
2753            Zhong Naixiong     $1 B          China   conglomerate  2674
2754           Zhou Wei family     $1 B          China       Software  2674

[2755 rows x 5 columns]

In [ ]:
data.describe()

Rank
count  2755.000000
mean   1345.663521
std     772.669811
min       1.000000
25%     680.000000
50%    1362.000000
75%    2035.000000
max    2674.000000

We see that Rank has the right data type i.e. int64 (which is the same as integer). However, NetWorth has a datatype of object, while it should be float or integer. Let's focus on this column first and fix its data type.

In [ ]:
# look at data for Networth column
data[['NetWorth', 'Country']]

NetWorth        Country
0      $177 B  United States
1      $151 B  United States
2      $150 B         France
3      $124 B  United States
4       $97 B  United States
...       ...            ...
2750     $1 B          China
2751     $1 B          China
2752     $1 B          China
2753     $1 B          China
2754     $1 B          China

[2755 rows x 2 columns]

In [ ]:
# let's look at the data type of one of the values.
first_value = data['NetWorth'].iloc[0]
print(first_value)
print(type(first_value))

$177 B
<class 'str'>


We see that it has $ and B to represent column values as currency. However, they also prevent it from being interpreted as numeric values. We can fix that by creating a new column that only contains the numeric content of this column. Currently, object data type is basically string so we need to cut down the string to only take the numeric value out of it.

In [ ]:
# let's test our logic by subsetting the first value to acquire the numeric content.
# Strings can be subset like arrays so here, if we use [0] we can access the first item of this string which is a $
print(first_value[0])
print(first_value[1:])

$
177 B


In [ ]:
# similarly, the last two values are a space and a B
first_value[-2:]

' B'

In [ ]:
# so combining both these indexing, we get the numeric content. However, that would still be string type
subset = first_value[1:-2]
print(subset)
print(type(subset))

177
<class 'str'>


In [ ]:
# so we need one more step to convert this data type into numeric. We use float().
# Since all values are explicitly rounded to Billion, however, just to be safe we use float().
# if a int() encounters a decimal value it throws an error however, float() can handle both decimal and rounded values
converted = float("117")
print(converted)
type(converted)

117.0


float

Perfect. However, there are 2700+ rows worth of values. Enter `apply()`. It is a function that lets you apply a particular function on each value of a dataframe row by row. Normally, you would do iteration using a `for` loop that would be very computationally intensive since we have 2700+ values. However, the `apply()` function does so with a single line of code; super efficiently. It takes an argument function either use-defined or pre-defined. We already figured out the logic on how to convert our data type. Let's convert that into a function and then use it with apply function.

In [ ]:
# defining a user-defined function called fix_networth(value)
# we already have the logic figured out, let's just copy the code here.

# $177 (object) ---> 177 (float)

def fix_networth(value):
    subset = value[1:-2]
    converted = float(subset)
    return converted


# this function takes one value at a time and fixes it to the format and data type we want. Let's merge that with apply()
new_column = data['NetWorth'].apply(fix_networth)
print(new_column)

In [86]:
# that's great! Let's store it in a new column called ['NetWorth ($B)']
data['NetWorth ($B)'] = new_column
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2755 entries, 0 to 2754
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Name           2755 non-null   object 
 1   NetWorth       2755 non-null   object 
 2   Country        2755 non-null   object 
 3   Source         2755 non-null   object 
 4   Rank           2755 non-null   int64  
 5   NetWorth ($B)  2755 non-null   float64
dtypes: float64(1), int64(1), object(4)
memory usage: 129.3+ KB
None


In [87]:
data.describe()

Rank  NetWorth ($B)
count  2755.000000    2755.000000
mean   1345.663521       4.749220
std     772.669811       9.615358
min       1.000000       1.000000
25%     680.000000       1.500000
50%    1362.000000       2.300000
75%    2035.000000       4.200000
max    2674.000000     177.000000

That's great! Now we are almost ready to get started to answering our questions. However, let's get rid of the old NetWorth column since its of no use to us anymore. For that we can use the .drop() function that takes the `columns=` argument in which you can specify what columns you want to remove from your dataset.

In [88]:
data

Name NetWorth               Country  \
0                                 Jeff Bezos   $177 B         United States   
1                                  Elon Musk   $151 B         United States   
2                   Bernard Arnault & family   $150 B                France   
3                                 Bill Gates   $124 B         United States   
4                            Mark Zuckerberg    $97 B         United States   
5                             Warren Buffett    $96 B         United States   
6                              Larry Ellison    $93 B         United States   
7                                 Larry Page  $91.5 B         United States   
8                                Sergey Brin    $89 B         United States   
9                              Mukesh Ambani  $84.5 B                 India   
10                            Amancio Ortega    $77 B                 Spain   
11     Francoise Bettencourt Meyers & family  $73.6 B                France   
12                            Zhong Shanshan  $68.9 B                 China   
13                             Steve Ballmer  $68.7 B         United States   
14                                Ma Huateng  $65.8 B                 China   
15                 Carlos Slim Helu & family  $62.8 B                Mexico   
16                              Alice Walton  $61.8 B         United States   
17                                Jim Walton  $60.2 B         United States   
18                                Rob Walton  $59.5 B         United States   
19                         Michael Bloomberg    $59 B         United States   
20                         Colin Zheng Huang  $55.3 B                 China   
21                           MacKenzie Scott    $53 B         United States   
22                            Daniel Gilbert  $51.9 B         United States   
23                     Gautam Adani & family  $50.5 B                 India   
24                      Phil Knight & family  $49.9 B         United States   
25                                   Jack Ma  $48.4 B                 China   
26                              Charles Koch  $46.4 B         United States   
27                       Julia Koch & family  $46.4 B         United States   
28                             Masayoshi Son  $45.4 B                 Japan   
29                              Michael Dell  $45.1 B         United States   
30                    Tadashi Yanai & family  $44.1 B                 Japan   
31                 François Pinault & family  $42.3 B                France   
32                    David Thomson & family  $41.8 B                Canada   
33         Beate Heister & Karl Albrecht Jr.  $39.2 B               Germany   
34                                  Wang Wei    $39 B                 China   
35                            Miriam Adelson  $38.2 B         United States   
36                              He Xiangjian  $37.7 B                 China   
37                            Dieter Schwarz  $36.9 B               Germany   
38                              Zhang Yiming  $35.6 B                 China   
39                          Giovanni Ferrero  $35.1 B                 Italy   
40                          Alain Wertheimer  $34.5 B                France   
41                         Gerard Wertheimer  $34.5 B                France   
42                               Li Ka-shing  $33.7 B             Hong Kong   
43                      Qin Yinglin & family  $33.5 B                 China   
44                          William Lei Ding    $33 B                 China   
45                             Len Blavatnik    $32 B         United States   
46                              Lee Shau Kee  $31.7 B             Hong Kong   
47                           Jacqueline Mars  $31.3 B         United States   
48                                 John Mars  $31.3 B         United States   
49                      Yang Huiyan & family  $29.6 B                 China   
50         

In [89]:
# drop old networth column
data = data.drop(columns=['NetWorth'])

In [ ]:
data2.drop(['Source'], axis=1)

In [90]:
data.head()

Name        Country         Source  Rank  NetWorth ($B)
0                Jeff Bezos  United States         Amazon     1          177.0
1                 Elon Musk  United States  Tesla, SpaceX     2          151.0
2  Bernard Arnault & family         France           LVMH     3          150.0
3                Bill Gates  United States      Microsoft     4          124.0
4           Mark Zuckerberg  United States       Facebook     5           97.0

In [ ]:
data.describe()

Rank  NetWorth ($B)
count  2755.000000    2755.000000
mean   1345.663521       4.749220
std     772.669811       9.615358
min       1.000000       1.000000
25%     680.000000       1.500000
50%    1362.000000       2.300000
75%    2035.000000       4.200000
max    2674.000000     177.000000

Looking great! Let's dig in!

### Answering Data Questions
Re-stating our questions here for ease.
1. How many billionaires are there in the list?
2. Who are the top 10 billionaire?
3. Who are the top billionaire by country of origin?
4. What country has the highest number of billionaires?
5. What country has the least number of billionaires?
6. What is the average and median networth of each billionaire?
7. Are there any outliers i.e. billionaires that have exceptionally high networth compared to others?
8. For each country, what is the cumulative networth of its billionaires?
9. What is the maximum and minimum networth of billionaires overall and by country?
10. Are there are any billionaires with the same source of income?

In [91]:
# Our first question, how many Billionaires in our dataset?
# We already know there are 2755 rows in the dataset. So that's how many billionaires there should be right?
# What if there are duplicates?
# We use the unique() function to get all unique values within a column. Let's check it out
data['Name'].unique()

array(['Jeff Bezos', 'Elon Musk', 'Bernard Arnault & family', ...,
       'Zhao Meiguang', 'Zhong Naixiong', 'Zhou Wei family'], dtype=object)

In [92]:
# this returns an array. Let's call .shape to find the shape
unique_values = data['Name'].unique()
unique_values.shape

(2752,)

In [ ]:
data[data["Name"].duplicated()]

Name Country                  Source  Rank  \
1625          Robert Miller  Canada  electronics components  1580   
1919  Wang Yanqing & family   China   carbon fiber products  1833   
1976                  Li Li   China         pharmaceuticals  1931   

      NetWorth ($B)  
1625            2.0  
1919            1.7  
1976            1.6

Looks like we had some duplicate values there. Back to data cleaning board.

#### Data Cleaning, Again

Now we have some duplicated values. We can use the duplicated() function to find out if there are any duplicates.

In [ ]:
data['Name'].duplicated()

That's a lot of Falses, but somewhere in there is a true. Now what we will do is use a new way to subset dataframe. We will use conditional **boolean masks** to retreive only those values that are marked duplicated by pandas. For that, we will use the following syntax

`df[conditional expression]`

That translates to:

`data[data['Name'].duplicated()]`

What this means is, return to us the original dataset where there are marked duplicates as we computed on the `Name` column.
the `conditional expression` can be applied on any column within the dataframe. We will see more examples of this later. For now, let's see our duplicated values.

In [ ]:
data[data['Name'].duplicated()]

Name Country                  Source  Rank  \
1625          Robert Miller  Canada  electronics components  1580   
1919  Wang Yanqing & family   China   carbon fiber products  1833   
1976                  Li Li   China         pharmaceuticals  1931   

      NetWorth ($B)  
1625            2.0  
1919            1.7  
1976            1.6

Looks like we have these three duplicated values with us. We can use the same conditional expression only this time we know the names we are looking for. So, now the expression looks something like `data['Name'] == 'Robert Miller'` for instance, where we want to retrieve all rows where the name is Robert Miller. We know it appears at least more than once since it has been marked duplicate. Let's have a look at all our duplicated names.

In [ ]:
data[(data['Name'] == 'Robert Miller')]

Name         Country                  Source  Rank  \
1331  Robert Miller  United Kingdom                  retail  1299   
1625  Robert Miller          Canada  electronics components  1580   

      NetWorth ($B)  
1331            2.4  
1625            2.0

In [ ]:
data[data['Name'] == 'Wang Yanqing & family']

Name Country                 Source  Rank  \
700   Wang Yanqing & family   China   electrical equipment   680   
1919  Wang Yanqing & family   China  carbon fiber products  1833   

      NetWorth ($B)  
700             4.2  
1919            1.7

In [ ]:
data[data['Name'] == 'Li Li']

Name Country           Source  Rank  NetWorth ($B)
693   Li Li   China       healthcare   680            4.2
1976  Li Li   China  pharmaceuticals  1931            1.6

Now there are two different scenarios here.
1. In case of Robert Miller, they are not the same. We notice this because their country, source and networth also differs. Which means they are two different people.
2. In case of the other two entries, both have the same country and name however, different source and networths.


For 1, we do a quick google search and find out that these are indeed two different people. We have an Canadian Robert Miller[https://en.wikipedia.org/wiki/Robert_Miller_(businessman)] and a American Robert Miller[https://en.wikipedia.org/wiki/Robert_Warren_Miller]

For 2, we also do some search and find Forbes mention their rank at 1976[https://www.forbes.com/profile/li-li-2/?sh=7caff2f216d9] for Li Li and at 680[https://www.forbes.com/profile/wang-yanqing/?sh=53096d9240a8] for Wang Yanqing & Family.


So our approach to handling duplicates here will be:
1. In case of Robert Miller, we will let this be.
2. In case of Wang Yanqing & Family, we will drop carbon fiber products entry.
3. In case of Li Li, we will drop healthcare entry.

In [93]:
# Since we now know that Name is not a unique list (we have two Robert Millers) so we have another problem to deal with.
# We will use drop() with argument for index where we will pass the index value of the rows we are about to drop.
fixed_data = data.drop(index=[693, 1919])

Now we know there is one duplicate name in Name column hence we can no longer use that for total count of billionaire.
We will however be able to use .shape since we have catered to the duplicates issue and now the total length correctly reflects the total number of billionaires.

In [94]:
# total billionaires
print(fixed_data.shape)

(2753, 5)


#### Additional Reference Material

If you need a quick understanding of the functions we used today in pandas, refer to this link and try it out on your own.
https://www.analyticsvidhya.com/blog/2021/04/20-must-known-pandas-function-for-exploratory-data-analysis-eda/

In the next section, we will continue with answering the rest of our questions. To quickly recap:

- We imported data into python using read_csv()
- We brainstormed and shortlisted questions to answer related to our dataset.
- We developed a checklist for data cleaning
    - handle data types
    - handle missing values
    - handle duplicates

Next, we move on to answering our data exploration questions.

# Part 2

We will continue with our previous dataset and attempt to answer our questions.

### Answering Data Questions
Re-stating our questions here for ease.
1. How many billionaires are there in the list?
2. Who are the top 10 billionaire?
3. Who are the top billionaire by country of origin?
4. What country has the highest number of billionaires?
5. What country has the least number of billionaires?
6. What is the average and median networth of each country's billionaire?
7. Are there any outliers i.e. billionaires that have exceptionally high networth compared to others?
8. For each country, what is the cumulative networth of its billionaires?
9. What is the maximum and minimum networth of billionaires overall and by country?
10. Are there are any billionaires with the same source of income?

Let's have a look at our data set that we have so far cleaned stored in `fixed_data` variable.

In [ ]:
fixed_data.head()

Name        Country         Source  Rank  NetWorth ($B)
0                Jeff Bezos  United States         Amazon     1          177.0
1                 Elon Musk  United States  Tesla, SpaceX     2          151.0
2  Bernard Arnault & family         France           LVMH     3          150.0
3                Bill Gates  United States      Microsoft     4          124.0
4           Mark Zuckerberg  United States       Facebook     5           97.0

Q1. How many billionaires are there in the list?

Answer: Since we have removed our duplicates and can use .shape to determine the number of rows hence our total billionaires.

In [ ]:
print('Number of Billionaires in our list: ', fixed_data.shape[0])

Number of Billionaires in our list:  2753


Q2. Who are the top 10 billionaires?

Answer: Now there are various ways to do this. You can use the `Rank` column and get values that `<= 10`. You could sort the values by `NetWorth` and then either use `head(n)` or index the first ten rows.

In [ ]:
# let's use the rank column approach.
top_ten = fixed_data[fixed_data['Rank'] <= 10]
top_ten['Name']

0                  Jeff Bezos
1                   Elon Musk
2    Bernard Arnault & family
3                  Bill Gates
4             Mark Zuckerberg
5              Warren Buffett
6               Larry Ellison
7                  Larry Page
8                 Sergey Brin
9               Mukesh Ambani
Name: Name, dtype: object

#### nlargest() and nsmallest() functions
Another easier way to retrieve top n (or bottom n) values based on a column is to use the `nlargest()` (or the `nsmallest()`) functions on the data frame.

In [ ]:
# alternate approach uses nlargest which retrieves all data for the top N values based on the specified column.
# the syntax is as follows:
# data.nlargest(n, columns=column_name)

alt_top_ten = fixed_data.nlargest(10, columns='NetWorth ($B)')
alt_top_ten['Name']

0                  Jeff Bezos
1                   Elon Musk
2    Bernard Arnault & family
3                  Bill Gates
4             Mark Zuckerberg
5              Warren Buffett
6               Larry Ellison
7                  Larry Page
8                 Sergey Brin
9               Mukesh Ambani
Name: Name, dtype: object

Try the other approaches on your own and see if you get the same results.

Q3. Who are the top billionaire by country of origin?

Answer: For this we will have to use the groupby function to group our data on set

In [101]:
# use group by to find the max value for each group
top_each_country = fixed_data.groupby('Country').apply(lambda x: x.nlargest(10, columns=['NetWorth ($B)']))
type(top_each_country)

pandas.core.frame.DataFrame

In [ ]:
top_each_country

In [ ]:
df3 = pd.DataFrame()

for country in fixed_data['Country'].unique():
  print(fixed_data.groupby('Country').get_group(country).nlargest(10, columns='NetWorth ($B)'))

Q4 and Q5. Which countries have the highest and the lowest number of billionaires?

Answer: We will group data by country and then use `.count()` function on the `Name` column.

In [104]:
# group data
grouped_by_country = fixed_data.groupby('Country')
# use .count() function on Name column
country_count = grouped_by_country['Name'].count()

# now we need to access the highest and the lowest values. Let's use nlargest and nsmallest()
print('Country with highest number of billionaires: \n', country_count.nlargest(2))
print('\n')
print('Country with lowest number of billionaires: \n', country_count.nsmallest(1))

Country with highest number of billionaires: 
 Country
United States    724
China            624
Name: Name, dtype: int64


Country with lowest number of billionaires: 
 Country
Algeria    1
Name: Name, dtype: int64


Q6. What is the average and median networth of each country's billionaires?

Answer: We will again group data by countries and then use mean and median on NetWorth columns

In [ ]:
# group data by country
grouped_by_country = fixed_data.groupby('Country')
# get average networth
mean_networth = grouped_by_country['NetWorth ($B)'].mean()
mean_networth.head(10)

Country
Algeria      4.800000
Argentina    3.060000
Australia    4.125000
Austria      5.225000
Belgium      4.200000
Brazil       3.256923
Canada       3.610937
Chile        4.744444
China        4.048077
Colombia     5.240000
Name: NetWorth ($B), dtype: float64

In [ ]:
# get median networth


In [ ]:
grouped_by_country = fixed_data.groupby('Country')
# use .count() function on Name column
country_mean = grouped_by_country['NetWorth ($B)'].mean()
print(country_mean)
#country_count.loc['Argentina']


# another way
#fixed_data[fixed_data['Country'] == 'Argentina']

In [105]:
# alternatively, we could use max() to get maximum number and idxmax() to get index of the maximum number
# Since in our grouped data the country name is the index hence it fulfills our requirement
print('Highest Billionaires are in', country_mean.idxmax(), 'which are', country_mean.max())

# same applies for min() and idxmin()
print('Least Billionaires are in', country_mean.idxmin(), 'which are', country_mean.min())

Highest Billionaires are in France which are 12.195238095238096
Least Billionaires are in Hungary which are 1.35


#### Food for Thought

You will notice that for some countries median and mean are the same while for others they may be different. Why do you think is that?

Q7 and Q8: Are there any outliers i.e. billionaires that have exceptionally high networth compared to others? For each country, what is the cumulative networth of its billionaires?

Answer: We can answer these questions together. First let's tackle Q8 i.e. cumulative networths by each country. This is relatively simple. AS we computed median and mean, we simply need to use the `sum()` function instead after grouping by `Country`.

In [ ]:
# group data by country
grouped_by_country = fixed_data.groupby('Country')

# generate total wealth by country
cum_wealth = grouped_by_country['NetWorth ($B)'].sum()

# print however you like since our question does not specify.
# Let's look at our top 10 wealthiest country based on billionaires.
cum_wealth.nlargest(10)

Country
United States     4398.4
China             2526.0
Germany            625.5
India              596.4
Russia             586.0
France             512.2
Hong Kong          448.2
Canada             231.1
Japan              225.4
United Kingdom     213.9
Name: NetWorth ($B), dtype: float64

Now back to Q7. Outliers are values that are generally far off from the general dataset, something that skews the mean.
For instance, in an undergrad first year class, you would normally expect students to be of age 17-19. However, if there is one new student aged 35 there, it would be regarded as an outlier.
We will learn more about these measures in a later section however for now this definition should suffice.

One way to detect outliers is by using quantiles. Quantiles is a way of dividing data into equally cut portions based on the data such that highest quantile contains highest values of the data and vice versa for lowest. The values of quantiles are represented in percentages between 0% to 100% (or between 0 to 1). Normally, if a dataset contains outliers, you can compute the mean and median and if the values are significantly different, then you can compute the quantile to see what the highest values are. You could also simply use nlargest() to find the largest values of the dataset but that requires a bit of trial and error to find the right value of n since the number of outliers is unknown.


For us, highest values are generally outliers so we will use 0.99 as our value for quantile which means show us the breakpoint where top 1% of the highest value are. Similarly, we can also use 0.95 for breakpoint for top 5% of the data and so on.

In [ ]:
# we need to find the quantile for entire dataset
top_1_pct = fixed_data['NetWorth ($B)'].quantile(0.01)
top_1_pct
# subset data to retrieve networth greater than these values
#fixed_data[fixed_data['NetWorth ($B)'] > top_1_pct]
#len(fixed_data[fixed_data['NetWorth ($B)'] <= top_1_pct])


1.0

We have a total of 2753 values and 1% of the data is 27 values. That is what we get. Notice how the top 4 values are significantly larger compared to the rest of the values for Net Worth. Now that we know that the top 4 are the outlier-ish values, we can use the `nlargest()` to obtain the top 4.

In [ ]:
# get the outlier values
print("Outliers in the dataset are as follows")
fixed_data.nlargest(4, columns='NetWorth ($B)')

Outliers in the dataset are as follows


Name        Country         Source  Rank  NetWorth ($B)
0                Jeff Bezos  United States         Amazon     1          177.0
1                 Elon Musk  United States  Tesla, SpaceX     2          151.0
2  Bernard Arnault & family         France           LVMH     3          150.0
3                Bill Gates  United States      Microsoft     4          124.0

Q9.What is the maximum and minimum networth of billionaires overall and by country?

Answer: For overall, we can simply call the `.max()` and `.min()` functions on our dataset. For country wise, we can do the same after grouping data by country.

Are there are any billionaires with the same source of income?

In [ ]:
# find max and min overall
print('Highest NetWorth of a Billionaire is', fixed_data['NetWorth ($B)'].max(), '$B')
print('Lowest NetWorth of a Billionaire is', fixed_data['NetWorth ($B)'].min(), '$B')

Highest NetWorth of a Billionaire is 177.0 $B
Lowest NetWorth of a Billionaire is 1.0 $B


In [ ]:
# find max and min by country
# group data
grouped_by_country = fixed_data.groupby('Country')
# get max billionaire wealth per country and sort countries
grouped_by_country['NetWorth ($B)'].max()

In [ ]:
grouped_by_country['NetWorth ($B)'].min()

Q10. Are there are any billionaires with the same source of income?

Answer: For this purpose, we can group data by source of income and then count billionaires in each section. After that sort by highest to lowest to see billionaires that have the same source of income and what exactly is that source.

In [ ]:
# group by source of income
group_by_source = fixed_data.groupby('Source')
# count the billionaires using name
source_count = group_by_source['Name'].count()
# sort by highest to lowest count
source_count.sort_values(ascending=False)

Source
real estate                  171
pharmaceuticals               99
investments                   90
diversified                   85
software                      62
                            ... 
hotels, energy                 1
hotels, diversified            1
supermarkets, investments      1
hospitals, health care         1
3D printing                    1
Name: Name, Length: 923, dtype: int64

Based on our calculations, the top 3 sources for becoming billionaires appear to be:
1. Real Estate
2. Pharmaceuticals
3. Investments


Do you belong to any of these fields?

### Try it Yourself

Figure out 1-2 more questions that you feel would provide more information on our dataset that we have not yet explored.

In [ ]:
### Start code here

In [ ]:
### end code here